RecoLocalCalo/EcalRecProducers/plugins/EcalPh2DigiToGPUProducer.cc

```diff
@@ -0,0 +1,105 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "HeterogeneousCore/CUDACore/interface/ScopedContext.h"
+#include "CUDADataFormats/EcalDigi/interface/DigisCollection.h"
+
+#include "DataFormats/EcalDigi/interface/EcalDigiCollections.h"
+#include "DataFormats/EcalDigi/interface/EcalDataFrame_Ph2.h"
+
+#include "DeclsForKernelsPh2.h"
+
+class EcalPh2DigiToGPUProducer : public edm::stream::EDProducer<edm::ExternalWork> {
+public:
+  explicit EcalPh2DigiToGPUProducer(const edm::ParameterSet& ps);
+  ~EcalPh2DigiToGPUProducer() override = default;
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  void acquire(edm::Event const& event, edm::EventSetup const& setup, edm::WaitingTaskWithArenaHolder holder) override;
+  void produce(edm::Event& evt, edm::EventSetup const& setup) override;
+
+private:
+  const edm::EDGetTokenT<EBDigiCollectionPh2> digiCollectionToken_;
+  const edm::EDPutTokenT<cms::cuda::Product<ecal::DigisCollection<calo::common::DevStoragePolicy>>>
+      digisCollectionToken_;
+  uint32_t n_;
+
+  ecal::DigisCollection<::calo::common::DevStoragePolicy> digis_;
+
+  cms::cuda::ContextState cudaState_;
+};
+
+void EcalPh2DigiToGPUProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("BarrelDigis", edm::InputTag("simEcalUnsuppressedDigis", ""));
+  desc.add<std::string>("digisLabelEB", "ebDigis");
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+EcalPh2DigiToGPUProducer::EcalPh2DigiToGPUProducer(const edm::ParameterSet& ps)
+    : digiCollectionToken_(consumes<EBDigiCollectionPh2>(ps.getParameter<edm::InputTag>("BarrelDigis"))),
+      digisCollectionToken_(produces<cms::cuda::Product<ecal::DigisCollection<calo::common::DevStoragePolicy>>>(
+          ps.getParameter<std::string>("digisLabelEB"))) {}
+
+void EcalPh2DigiToGPUProducer::acquire(edm::Event const& event,
+                                       edm::EventSetup const& setup,
+                                       edm::WaitingTaskWithArenaHolder holder) {
+  cms::cuda::ScopedContextAcquire ctx{event.streamID(), std::move(holder), cudaState_};
+
+  //input data from event
+  const auto& pdigis = event.get(digiCollectionToken_);
+
+  n_ = pdigis.size();
+
+  //allocate device pointers for output
+  digis_.ids = cms::cuda::make_device_unique<uint32_t[]>(n_, ctx.stream());
+  digis_.data = cms::cuda::make_device_unique<uint16_t[]>(n_ * EcalDataFrame_Ph2::MAXSAMPLES, ctx.stream());
+
+  //allocate host vectors for holding product data and id vectors
+
+  std::vector<uint32_t, cms::cuda::HostAllocator<uint32_t>> idstmp;
+  std::vector<uint16_t, cms::cuda::HostAllocator<uint16_t>> datatmp;
```

Comment:
The use of `std::vector<T, cms::cuda::HostAllocator<T>>` is unsafe here without synchronizing the `ctx.stream()` before `produce()` returns, because these objects may have been destructed (and the memory used for something else) before the asynchronous `cudaMemcpyAsync()` actually perform the copy.

It seems to me that only `std::vector::data()` is being used for these, in which case it would be easiest to use
```suggestion
  auto idstmp = cms::cuda::make_host_unique<uint32_t[]>(n_, ctx.stream());
  auto datatmp = cms::cuda::make_host_unique<uint16_t[]>(n_ * EcalDataFrame_Ph2::MAXSAMPLES, ctx.stream());
```
which keeps the host memory blocks alive until the asynchronous work queued into `ctx.stream()` has completed.